src/effects/cold_breath.c

```diff
@@ -23,4 +113,63 @@ void cold_breath_render(EffectInstance* effect) {
     retTask->renderMode |= RENDER_TASK_FLAG_2;
 }
 
-INCLUDE_ASM(s32, "effects/cold_breath", cold_breath_appendGfx);
+void cold_breath_appendGfx(EffectInstance* effect) {
+    ColdBreathFXData* data = effect->data.coldBreath;
+    Camera* camera = &gCameras[gCurrentCameraID];
+    s32 unk14 = data->unk_14;
+    s32 unk24 = data->unk_24;
+    s32 unk00 = data->unk_00;
+    Matrix4f sp20;
+    Matrix4f sp60;
+    s32 unkIndex;
+    f32 temp1;
+    f32 temp2;
+    f32 temp3;
+    f32 temp_f32;
+    s32 temp_s32;
+    s32 envAlpha;
+    s32 cond;
+
+    gDPPipeSync(gMasterGfxPos++);
+    gSPSegment(gMasterGfxPos++, 0x09, VIRTUAL_TO_PHYSICAL(effect->graphics->data));
+
+    shim_guTranslateF(sp20, data->unk_04, data->unk_08, data->unk_0C);
+    shim_guScaleF(sp60, data->unk_40, data->unk_40, data->unk_40);
+    shim_guMtxCatF(sp60, sp20, sp20);
+    shim_guMtxF2L(sp20, &gDisplayContext->matrixStack[gMatrixListPos]);
+
+    gSPMatrix(gMasterGfxPos++, &gDisplayContext->matrixStack[gMatrixListPos++], G_MTX_PUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPMatrix(gMasterGfxPos++, camera->unkMatrix, G_MTX_NOPUSH | G_MTX_MUL | G_MTX_MODELVIEW);
+    gDPSetPrimColor(gMasterGfxPos++, 0, 0, data->unk_18, data->unk_1C, data->unk_20, 0.5 * unk24);
+    gDPSetAlphaDither(gMasterGfxPos++, G_AD_NOISE);
+    gSPDisplayList(gMasterGfxPos++, D_E00DE84C[unk00]);
+
+    unkIndex = (unk14 < 30 ? unk14 : 29) + unk00 * 30;
+    temp1 = D_E00DE96C[unkIndex];
+    temp2 = D_E00DE8B4[unkIndex];
+    temp3 = D_E00DE858[unkIndex];
+
+    shim_guPositionF(sp20, 0.0f, 0.0f, -temp1, temp2 * 0.01f, temp3, 0.0f, 0.0f);
+    shim_guMtxF2L(sp20, &gDisplayContext->matrixStack[gMatrixListPos]);
+    gSPMatrix(gMasterGfxPos++, &gDisplayContext->matrixStack[gMatrixListPos++], G_MTX_PUSH | G_MTX_MUL | G_MTX_MODELVIEW);
+
+    temp_f32 = D_E00DE910[unkIndex] * 5.0f / 256.0f;
+    temp_s32 = temp_f32;
+    envAlpha = (temp_f32 - temp_s32) * 256.0f;
+
+    cond = temp_s32 >= 4;
```

Comment:
This is just terrible, but it refuses to match without it.

Reply:
you may want to try `temp_s32 == 4 || temp_s32 == 5` ? ohhhh or this is a switch on temp_s32 where the only two cases are 4 and 5. I would try that as well. you may be able to have 4 fall through to 5, or copy/paste the contents into both cases

Reply:
I'm not sure if we can check `temp_s32 < something` in a switch block. What we have is basically:
```c
if (...) {
  if (...) {
    // executed if temp_s32 == 4
  } else {
    // executed if temp_s32 < 4
  }
  // executed if temp_s32 < 5
}
```
Not sure if this helps, but for some reason the assembly code first calculates the `temp_s32 < 4` condition and then negates it with a xor operation ([lines 3f8 and 3fc](https://decomp.me/scratch/iRTJc)), a pattern I haven't seen in other functions. I tried moving the conditions around in couple of ways but to no avail.

Reply:
Oh I see. I think it's probably fine to keep it the way it is..though maybe wiseguy will be able to help (he's a gfx master). let me ask him

Reply:
> you may want to try `temp_s32 == 4 || temp_s32 == 5` ? ohhhh or this is a switch on temp_s32 where the only two cases are 4 and 5. I would try that as well. you may be able to have 4 fall through to 5, or copy/paste the contents into both cases

Actually, I think you may be right about the switch. If `temp_s32` was unsigned, it would make sense. Something like:
```cs
switch (temp_s32) {
    case 4:
        // executed if temp_s32 == 4
        // executed if temp_s32 < 5
        break;
    case 0:
    case 1:
    case 2:
    case 3:
        // executed if temp_s32 < 4;
        // executed if temp_s32 < 5;
}
```

I'm not sure if I can make it an unsigned var, though. Will try a couple of things.

Reply:
I'm not too worried about this, really. it's not terrible to read, and we can always revisit it later. if you try a bit and can't get a switch to work, np

Reply:
Yeah, couldn't got it to work :/ PR is ready to merge then.

Reply:
kk, we just need alex to look at it